Fix for https://github.com/AxonFramework/AxonFramework/issues/2341

heutelbeck opened the pull request:
Extended the type checking. Please check if there are other potential type issues going on.

smcvb commented on messaging/src/main/java/org/axonframework/queryhandling/SimpleQueryUpdateEmitter.java:
```diff
@@ -176,12 +180,24 @@ private <U> void doEmit(Predicate<SubscriptionQueryMessage<?, ?, U>> filter,
+			}
+			if(sqm.getUpdateResponseType() instanceof OptionalResponseType) {
+				return Optional.class.isAssignableFrom(payloadType);
+			}			
```
I would expect a third caveat for the `PublisherResponseType`.

smcvb replied:
And I'd expect a test case for this, of course.

smcvb reviewed (changes requested):
So, referring to your comment on issue #2341:

> So I figure the added type check introduced the bug. 
> As far as I understand, the type-checking mechanisms of the ResponseType classes are primarily for checking method parameters and not for updates at runtime.
> 
> So I propose to extend the type checking in the UpdateEmitter. 
> Due to type erasure, you can only check if the object is an Iterable etc. but not if the content of the Iterable matches the type. 
> The trivial "matches" approach via the ResponseType does not work reliably for multiple instances and for Optionals.

The `ResponseType#matches` method is used to deduce whether a Query Handling method can handle a specific `QueryMessage`. Pre-AF-4.6.0 also used it when finding the Query Handling method, but we've adjusted that for a handler priority approach for "most matching" with the `ResponseType#matchRank` method.

That said, we never used it for filter the updates from a `QueryUpdateEmitter`, nor was it intended too.
Sorry for not making that specific in your previous pull request. In all honesty, this point slipped my mind.

So, thanks for the pull request once again! I've added comments to specify we should also deal with the `PublisherResponseType`. Although I don't necessarily like the exceptional cases like this, I think it's justified in Axon's current lifecycle.

heutelbeck commented:
Yes, I only figured out the design intent to focus on inspecting methods for handlers only after adding the filter.  I added the type check for the ```Publisher``` and added some tests to verify the type check works.

smcvb reviewed (approved):
My concerns have been addressed, hence I'm approving this pull request.

Morlack reviewed (approved):
Looks good to me :+1:
